### PR TITLE
Lock storage while processing payments

### PR DESCRIPTION
**Pull request by vsbogd**
This is to fix issue https://github.com/singnet/snet-daemon/issues/109

What is done:
- modules separated from payment handler implementation:
  - payment channel api  to use from external modules
  - PaymentChannelService to return channel state, apply payment on channel and claim payments
  - ChannelPaymentValidator to validate payment using payment channel state
  - BlockchainChannelReader to read channel state from blockchain
- payment handler, state service and claim command are implemented using PaymentChannelService, so different PaymentChannelService implementations can use different lock logic
- lockingPaymentChannelService is PaymentChannelService implementation to lock payment channel on start of payment transaction and unlock after payment is applied
- if client tries to make another payment when channel is locked then FailedPrecondition error code is returned to client with appropriate error message
- locking logic is implemented using etcd storage with separate prefix, if storage contains "locked" string by channel name key then channel is locked

### CODE REVIEW

File: escrow/escrow.go [hunk 1]
```diff
@@ -1,267 +1,209 @@
...
-func (h *escrowPaymentHandler) Type() (typ string) {
-	return EscrowPaymentType
+func (claim *claimImpl) Payment() *Payment {
+	return claim.payment
 }
 
-func (h *escrowPaymentHandler) Payment(context *handler.GrpcStreamContext) (payment handler.Payment, err *status.Status) {
-	channelID, err := handler.GetBigInt(context.MD, PaymentChannelIDHeader)
+func (claim *claimImpl) Finish() (err error) {
+	err = claim.lock.Unlock()
```
Comment (stellarspot) [inline escrow/escrow.go]: It was discussed to check where this lock is really necessary.

File: escrow/escrow.go [hunk 2]
```diff
@@ -1,267 +1,209 @@
...
 	if !ok {
-		log.Warn("Payment channel not found")
-		return nil, status.Newf(codes.InvalidArgument, "payment channel \"%v\" not found", channelKey)
-	}
-
-	amount, err := handler.GetBigInt(context.MD, PaymentChannelAmountHeader)
+		return nil, fmt.Errorf("another transaction on channel: %v is in progress", key)
+	}
+	defer func(lock Lock) {
+		if err != nil {
```
Comment (stellarspot) [inline escrow/escrow.go]: It was discussed to check if the lock needs to be unlocked in any case.

File: config/config.go [hunk 3]
```diff
@@ -245,3 +245,9 @@ func LogConfig() {
 		}
 	}
 }
+
+func GetBigIntFromViper(config *viper.Viper, key string) (value *big.Int, err error) {
+	value = &big.Int{}
```
[the remark this answers is not in the document]
Comment (vsbogd) [reply config/config.go]: I would say it is two different cases:
- GetBigIntFromViper - we don't know actual value and constructing empty value to unmarshall it
- big.NewInt(0) is used to set 0 to the struct member

File: snetd/cmd/claim.go [hunk 4]
```diff
@@ -58,8 +56,8 @@ func newClaimCommand(cmd *cobra.Command, args []string, components *Components)
 	}
 
 	command = &claimCommand{
-		storage:    components.PaymentChannelStorage(),
-		blockchain: components.Blockchain(),
+		channelService: components.PaymentChannelService(),
+		blockchain:     components.Blockchain(),
```
[the remark this answers is not in the document]
Comment (vsbogd) [reply snetd/cmd/claim.go]: I would do this later, as separate PR

File: escrow/lock.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	locked   = "locked"
+	unlocked = "unlocked"
+)
+
+func (locker *etcdLocker) Lock(name string) (lock Lock, ok bool, err error) {
+	_, ok, err = locker.storage.Get(name)
+	if err != nil {
+		return
+	}
+	if ok {
```
[the remark this answers is not in the document]
Comment (vsbogd) [reply escrow/lock.go]: Agree, fixed

File: escrow/escrow.go (hunk 1 above)
Comment (vsbogd) [reply escrow/escrow.go]: Thank you for spotting it! Fixed it and https://github.com/singnet/snet-daemon/pull/128#discussion_r233364630 in https://github.com/singnet/snet-daemon/pull/128/commits/8f09533a7e46be8bb0670774642346b0d8e6cb59

**Pull request merged**